torch/utils/cpp_extension.py

```diff
@@ -1496,16 +1529,23 @@ def _get_cuda_arch_flags(cflags: Optional[List[str]] = None) -> List[str]:
 
     # If not given, determine what's best for the GPU / CUDA version that can be found
     if not _arch_list:
-        capability = torch.cuda.get_device_capability()
-        supported_sm = [int(arch.split('_')[1])
-                        for arch in torch.cuda.get_arch_list() if 'sm_' in arch]
-        max_supported_sm = max((sm // 10, sm % 10) for sm in supported_sm)
-        # Capability of the device may be higher than what's supported by the user's
-        # NVCC, causing compilation error. User's NVCC is expected to match the one
-        # used to build pytorch, so we use the maximum supported capability of pytorch
-        # to clamp the capability.
-        capability = min(max_supported_sm, capability)
-        arch_list = [f'{capability[0]}.{capability[1]}']
+        arch_list = []
+        # the assumption is that the extension should run on any of the currently visible cards,
+        # which could be of different types - therefore all archs for visible cards should be included
+        for i in range(torch.cuda.device_count()):
+            capability = torch.cuda.get_device_capability(i)
+            supported_sm = [int(arch.split('_')[1])
+                            for arch in torch.cuda.get_arch_list() if 'sm_' in arch]
+            max_supported_sm = max((sm // 10, sm % 10) for sm in supported_sm)
+            # Capability of the device may be higher than what's supported by the user's
+            # NVCC, causing compilation error. User's NVCC is expected to match the one
+            # used to build pytorch, so we use the maximum supported capability of pytorch
+            # to clamp the capability.
+            capability = min(max_supported_sm, capability)
```

Comment:
IMO the capability of the device is higher than what's supported by the user's NVCC, then the user has no business using that NVCC on that system and needs to upgrade their cuda toolkit and driver, and we should throw an error telling them so.

@ngimel what do you think?

If we do want to enable using an older toolkit for GPUs with CCs newer than the toolkit supports, we should make sure to tack `+PTX` onto the highest `capability` we end up compiling for.

Reply:
I think erroring out is too strict. E.g. now pytorch binaries are available only with cuda 11.0 which does not support 8.6 (30xx cards), and our recommendation has been to compile extensions with the same nvcc as libtorch. For most extensions having an absolute latest compiler is not a requirement, and will provide marginal perf gains at best.

Reply:
> our recommendation has been to compile extensions with the same nvcc as libtorch

So, then it shouldn't check the cc of the visible cards and instead  `torch.cuda.get_arch_list()` by default?

Reply:
But we should still tack `+PTX` onto the highest CC we compile for in this case, right?

Reply:
Yes, we should do +PTX.

Reply:
done